src/core.js

```diff
@@ -0,0 +1,655 @@
+     * @property {DispatchEvent} dispatchEvent - Dispatch an event to this input.
+     */
+
+    /**
+     * @typedef MIDIOutput
+     * @type {MIDIPort}
+     * @property {Function} send - A function that handles requests to send messages to this port.
+     * @property {Function} clear - A function that handles requests to clear messages in progress being sent by this port.
+     */
+
+    /**
+     * An array of data representing a MIDI message, for full details, see:
+     * https://www.midi.org/specifications-old/item/table-1-summary-of-midi-message
+     *
+     * @typedef MIDIMessageData
+     * @type {Uint8Array}
+     *
+     */
+    youme.requestAccess = function (sysex, software, onAccessGranted, onError) {
+        if (!navigator.requestMIDIAccess) {
+            var msg = "The Web MIDI API is not available. You may need to enable it in your browser's settings.";
+            fluid.log(fluid.logLevel.WARN, msg);
+            onError(msg);
+            return;
+        }
+
+        var p = navigator.requestMIDIAccess({
+            sysex: sysex,
+            software: software
+        });
+
+        p.then(onAccessGranted, onError);
+    };
+
+    youme.getPorts = function (access) {
+        var ports = {};
+
+        youme.collectPorts("inputs", access, ports);
+        youme.collectPorts("outputs", access, ports);
+
+        return ports;
+    };
+
+    youme.requestPorts = function (success, error, software) {
+        function wrappedSuccess(access) {
+            var ports = youme.getPorts(access);
+            success(ports);
+        }
+
+        youme.requestAccess(false, software, wrappedSuccess, error);
+    };
+
+    youme.portsToViews = function (portsArray) {
+        return fluid.transform(portsArray, function (port) {
+            return fluid.filterKeys(port, ["id", "name", "manufacturer", "state", "connection"]);
+        });
+    };
+
+    youme.prettyPrintPorts = function (ports) {
+        return fluid.prettyPrintJSON({
+            inputs: youme.portsToViews(ports.inputs),
+            outputs: youme.portsToViews(ports.outputs)
+        });
+    };
+
+    youme.logPorts = function (software) {
+        function success(ports) {
+            var printed = youme.prettyPrintPorts(ports);
+            console.log(printed); // eslint-disable-line no-console
+        }
+
+        function error(err) {
+            console.log(err); // eslint-disable-line no-console
+        }
+
+        youme.requestPorts(success, error, software);
+    };
+
+    /**
+     * Convert an object that presents a map of `inputs` and `outputs` into an array of MIDIPort objects.
+     *
+     * @param {"inputs"|"outputs"} type - The type of inputs, only "inputs" and "outputs" are supported.
+     * @param {MIDIAccess} access - The `MIDIAccess` object provided by the WebMIDI API.
+     * @param {Map<MIDIPort>} ports - A read-only map-like structure containing the available ports for `type`.
```

Comment:
Nicely cleaned up, however, remember to remind the user that this "read-only" structure is actually modified!
Since the only caller of collectPorts is getPorts, and this function is now a one-liner, why not abolish it and just write the spreads directly in getPorts?

Reply:
Done.